project/driver_controls/src/input_interrupt.c

```diff
@@ -0,0 +1,149 @@
+#include "adc.h"
+#include "input_interrupt.h"
+#include "event_queue.h"
+#include "stm32f0xx.h"
+
+#define NUM_INPUT_EVENTS 16
+
+#define COAST_THRESHOLD 1000
+#define DRIVE_THRESHOLD 3000
+
+typedef void (*DriverInput)(GPIOAddress address, Event *e);
+
+typedef struct AnalogStatus {
+  InputEvent pedal;
+} AnalogStatus;
+
+static void prv_direction_selector(GPIOAddress address, Event *e);
+static void prv_turn_signal(GPIOAddress address, Event *e);
+static void prv_mechanical_brake(GPIOAddress address, Event *e);
+static void prv_event(GPIOAddress address, Event *e);
+
+static AnalogStatus s_analog_status;
+
+// Possibly replace with uint16_t later on
+static bool s_input_status[16];
+
+static const DriverInput s_driver_inputs[] = {
+  prv_event,
+  NULL,
+  prv_direction_selector,
+  prv_direction_selector,
+  prv_event,
+  prv_event,
+  prv_event,
+  prv_turn_signal,
+  prv_turn_signal,
+  prv_event,
+  prv_mechanical_brake
+};
+
+static InputEvent s_event_lookup[NUM_INPUT_EVENTS] = {
+  INPUT_EVENT_POWER,
+  INPUT_EVENT_GAS_BRAKE,
+  INPUT_EVENT_GAS_COAST,
+  INPUT_EVENT_GAS_PRESSED,
+  INPUT_EVENT_CRUISE_CONTROL,
+  INPUT_EVENT_CRUISE_CONTROL_INC,
+  INPUT_EVENT_CRUISE_CONTROL_DEC,
+  INPUT_EVENT_HAZARD_LIGHT
+};
+
+static prv_update_status(GPIOAddress address, GPIOState key_pressed) {
+  s_input_status[address.pin] = key_pressed;
+}
+
+static void prv_direction_selector(GPIOAddress address, Event *e) {
+  uint8_t status = (s_input_status[3] << 1) | s_input_status[2];
```

Comment:
This is pretty magical. It would be nice if you used macros and enums instead of indexing into random arrays.